contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,491 @@
+name of each service instance will be same as the stateful instance name assigned to the
+stateful replicas. This will be automatically created by the federated statefulset controller,
+on demand, proposed to be controlled by an annotation on the statefulset. If the annotation
+specifies so (users choice), the federation controller will create the additional service of
+type &#39;LoadBalancer&#39; against each pod replica. The key name for the annotation will be:
+```federation.kubernetes.io/statefulset-lb-needed```
+The values could be ```true``` or ```false```. The default value will be ```false```, considering
+that the external load balancer is an expensive resource.
+
+The service name for the LB service, one each against the stateful pod will be the same name as
+the stateful pod unique name. Also note that these particular LB services, although federated,
+will ideally resolve to 1 IP per service. The LB services will be created with proper label
+selectors to ensure the network traffic can reach the designated pods. This would also mean
+that each stateful pod will need to be updated with an additional unique label selector which
+can be matched with by the individual LB services. The suggestion is to use
+```federated-statefulset: <stateful replica unique name>``` as the label selector.
+
+The details of creation and handling the LB services are elaborated in section **Handling
+additional LB Services** later in this doc.
+
+The trick for individual pods being able to discover this additional dns identity through the LB
+service is the federated controller first creating SRV records for each stateful pod in the
+federation dns server against the governing service domain
+```store.mynamespace.myfederation.svc.example.com``` and then later updating an A record of the IP
+resolved for the specific LB service (for example ```store-c1-0.mynamespace.myfederation.svc.example.com```)
+against each of these SRV records.
+A peer pod in a stateful set then can discover its federated peers by doing a dns query against
+the governing service domain ```store.mynamespace.myfederation.svc.example.com```.
+It should be noted that as of now the federation service controller ignores service records which
+are not of type Loadbalancer, and does not update them in the federation dns server. This behaviour
+will be aptly modified to ensure that the dns records against the headless services are also updated
+in the federated dns server if the need be. The federation service controller will determine this based
+on the user annotation (```federation.kubernetes.io/statefulset-lb-needed```) with value ```true``` present
+on the federated headless service. Please not that this is the same annotation which is used on the
+federated statefulset to determine, if this statefulset needs LB services or not. The federation
+statefulset controller will update this on the federated headless service while creating it, or if
```

Comment:
Updated to -> "the stateful set controller will synchronize the headless service with the annotation value"